staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -307,19 +473,166 @@ func (q *delayingType) waitingLoop() {
 	}
 }
 
-// insert adds the entry to the priority queue, or updates the readyAt if it already exists in the queue
+// insert adds the entry to the 'waiting' queue, or updates the readyAt if it already exists in the queue
+// It returns a boolean to indicate that the entry was inserted or not.
 func insert(q *waitForPriorityQueue, knownEntries map[t]*waitFor, entry *waitFor) {
-	// if the entry already exists, update the time only if it would cause the item to be queued sooner
+	// if the entry already exists, update the time according to the entry.options.WhenWaiting
 	existing, exists := knownEntries[entry.data]
 	if exists {
-		if existing.readyAt.After(entry.readyAt) {
+		var replace bool
+		switch entry.options.WhenWaiting {
+		case TakeShorter:
+			replace = existing.readyAt.After(entry.readyAt)
+		case TakeLonger:
+			replace = existing.readyAt.Before(entry.readyAt)
+		case TakeIncoming:
+			replace = !existing.readyAt.Equal(entry.readyAt)
+		}
+		if replace {
```

Comment:
Not sure I get you there?  All values for AlreadyWaitingBehaviour are handled (TakeShorter, TakeLonger and TakeIncoming are handled in insert(), and TakeExisting() is handled in processArrivingWaitForEntry() because it can deal with it quicker there without requiring the call to insert().

I want the client to be able to control all behaviours through AddWithOptions() with the idea that eventually AddAfter could be perhaps retired.

Reply:
TakeShorter is still the default behaviour: TakeShorter AlreadyWaitingBehaviour = iota

Reply:
Ah, you are right, I made a mistake when reviewing it. 

Nevertheless would you mind still explicitly passing `TakeShorter` into the original `AddAfter` implementation  to emphasize what we want to achieve there? I think it would help to document the function behaviour better even when it is the default behaviour.

Reply:
Have done